File: client/app/src/main/java/fr/coopuniverse/api/pokeapi/activity/httpRequestManager/CallBackGenerator.java
package fr.coopuniverse.api.pokeapi.activity.httpRequestManager;

import android.os.AsyncTask;
import android.util.Log;

import java.io.IOException;
import java.util.List;

import fr.coopuniverse.api.pokeapi.activity.InventoryActivity;
import retrofit2.Call;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class CallBackGenerator extends AsyncTask<Void, Void, Void> {

    public String url;
    private String idUser;
    private String idUserTwo;
    private List<Card> cards;
    private Card cardUserOne;
    private Card cardUserTwo;
    private String answser;
    private String value;
    private Reponse reponse;
    private String action;

    public CallBackGenerator setUrl(String url) {
        this.url = url;
        return this;
    }

    public CallBackGenerator setIdUser(String idUser) {
        this.idUser = idUser;
        return this;
    }

    public CallBackGenerator setIdUserTwo(String idUserTwo) {
        this.idUserTwo = idUserTwo;
        return this;
    }

    public CallBackGenerator setCards(List<Card> cards) {
        this.cards = cards;
        return this;
    }

    public CallBackGenerator setCardUserOne(Card cardUserOne) {
        this.cardUserOne = cardUserOne;
        return this;
    }

    public CallBackGenerator setCardUserTwo(Card cardUserTwo) {
        this.cardUserTwo = cardUserTwo;
        return this;
    }

    public CallBackGenerator setAnswser(String answser) {
        this.answser = answser;
        return this;
    }

    public CallBackGenerator setValue(String value) {
        this.value = value;
        return this;
    }
    public CallBackGenerator setAction(String action) {
        this.action = action;
        return this;
    }

    public void generateCallBack() {
        Retrofit retrofit = new Retrofit.Builder().baseUrl(url).addConverterFactory(GsonConverterFactory.create()).build();
        CoopUniverseService service = retrofit.create(CoopUniverseService.class);
        Call<Reponse> rep;
        switch (action)
        {
            case("GetOneMoney"):
                rep = service.GetOneMoney(this.idUser);
                break;
            case("GetAllMoney"):
                rep = service.GetAllMoney();
                break;
            case("SetOneMoney"):
                rep = service.SetOneMoney(this.idUser, this.value);
                break;
            case("SetAllMoney"):
                rep = service.SetAllMoney(this.value);
                break;
            case("GetOneUser"):
                rep = service.GetOneUser(this.idUser);
                break;
            case("SetOneUser"):
                rep = service.SetOneUser(this.idUser,this.value);
                break;
            case("GetAllUser"):
                rep = service.GetAllUser();
                break;
            case("SetAllUser"):
                rep = service.SetAllUser(this.value);
                break;
            case("GetOneCardByUserId"):
                rep = service.GetOneCardByUserId(this.idUser);
                break;
            case("GetAllCard"):
                rep = service.GetAllCard();
                break;
            case("SetOneCardByUserId"):
                rep = service.SetOneCardByUserId(this.idUser,this.cardUserOne.toString());
                break;
            case("DeleteOneCardByUserid"):
                rep = service.DeleteOneCardByUserid(this.idUser);
                break;
            case("ExchangeCards"):
                rep = service.ExchangeCards(this.idUser,this.idUserTwo,this.cardUserOne.toString(),this.cardUserTwo.toString());
                break;
            case("GetRandomCard"):
                rep = service.GetRandomCard(this.idUser);
                break;
            case("MeltCards"):
                rep = service.MeltCards(this.idUser,this.cards.toString());
                break;
            case("CraftOneCard"):
                rep = service.CraftOneCard(this.idUser, this.cards.toString());
                break;
                default:
            case("getQuestion"):
                rep = service.getQuestion(this.idUser);
                break;
            case("setAnswer"):
                rep = service.setAnswer(this.idUser, this.answser);
                break;
        }

        try {
            reponse = rep.execute().body();

        } catch (IOException e)
        {
            e.printStackTrace();
        }


    }

    @Override
    protected Void doInBackground(Void... voids)
    {
        generateCallBack();
        return null;
    }

    protected void onPostExecute(Void result)
    {
        if(reponse != null)
            InventoryActivity.respond.setText(reponse.toString());


    }
}

File: client_kotlin/app/src/main/java/fr/coopuniverse/api/pokeapi/activity/httpRequestManager/CallBackGenerator.kt
package fr.coopuniverse.api.pokeapi.activity.httpRequestManager

import android.os.AsyncTask
import android.util.Log
import android.widget.TextView
import fr.coopuniverse.api.pokeapi.activity.activity.CallBackDisplay

import java.io.IOException

import retrofit2.Call

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class CallBackGenerator(
        var callback: CallBackDisplay,
        var isActivateCallBack: Boolean? = false,
        var typeFilter: String? = null,
        var valueFilter: String? = null,
        var url: String? = null,
        var idUser: String? = null,
        var idUserTwo: String? = null,
        var idCard: String? = null,
        var idCardOne: String? = null,
        var idCardTwo: String? = null,
        var idCardThree: String? = null,
        var cardUserOne: String? = null,
        var cardUserTwo: String? = null,
        var answser: String? = null,
        var value: String? = null,
        var action: String? = null,
        var key: String? = null,
        var login: String? = null,
        var pass: String? = null)
    : AsyncTask<TextView, Void, Reponse>() {


    fun generateCallBack(): Reponse {

        Log.d("Chaton",""+pass + " " + login)
        var reponse: Reponse?
        val retrofit = Retrofit.Builder().baseUrl(url).addConverterFactory(GsonConverterFactory.create()).build()
        val service = retrofit.create(CoopUniverseService::class.java)
        val rep: Call<Reponse>
        when (action) {
            "GetOneMoney" -> rep = service.GetOneMoney(idUser!!)
            "GetAllMoney" -> rep = service.GetAllMoney()
            "GetOneUser" -> rep = service.GetOneUser(idUser!!)
            "GetAllUser" -> rep = service.GetAllUser()
            "GetCardByUserId" -> rep = service.GetCardByUserId(idUser!!)
            "GetAllCard" -> rep = service.GetAllCard()
            "GetRandomCard" -> rep = service.GetRandomCard()
            "GetAllParameter" -> rep = service.GetAllParameter()
            "GetQuestion" -> rep = service.GetQuestion()
            "GetCardByFilter" -> rep = service.GetCardByFilter(typeFilter!!,valueFilter!!)
            "SetOneCard" -> rep = service.SetOneCard(idUser!!,idCard!!)
            "SetOneMoney" -> rep = service.SetOneMoney(idUser!!, value!!)
            "SetAnswer" -> rep = service.SetAnswer(idUser!!, answser!!)
            "ExchangeCards" -> rep = service.ExchangeCards(idUser!!, idUserTwo!!, cardUserOne!!.toString(), cardUserTwo!!.toString())
            "MeltCards" -> rep = service.MeltCards(idUser!!,idCard!!)
            "CraftOneCard" -> rep = service.CraftOneCard(idUser!!,idCardOne!!,idCardTwo!!,idCardThree!!)
            "Connect" -> rep = service.SimpleLogin(login!!, pass!!)
            "ConnectFacebook" -> rep = service.FacebookLogin(key!!)
            "ConnectGoogle" -> rep = service.GoogleLogin(key!!)
            "Register" -> rep = service.SimpleRegister(login!!, pass!!)
            "RegisterFacebook" -> rep = service.FacebookRegister(key!!)
            "RegisterGoogle" -> rep = service.GoogleRegister(key!!)
            else -> rep = service.GetAllCard()
        }

        try {

            reponse = rep.execute().body()
            return reponse

        } catch (e: IOException) {
            e.printStackTrace()

        }
        return Reponse()

    }

    override fun doInBackground(vararg TextViews: TextView?): Reponse? {

        return generateCallBack()
    }

    override fun onPostExecute(result: Reponse)
    {

        if(isActivateCallBack == true)
        {
            callback.display(result)
        }


    }
}
